### PR TITLE
feat: use the revision from Cargo.lock

### LINKED ISSUE #109: No rev specified
I'm getting an error which I don't understand. As far as I can tell, I have specified revisions for all of the dependencies, but it errors with `
No 'rev', 'tag' or 'branch' specified`.

```json
{
    "naersk": {
        "branch": "master",
        "description": "Build rust crates in Nix. No configuration, no code generation, no IFD. Sandbox friendly.",
        "homepage": "",
        "owner": "nmattia",
        "repo": "naersk",
        "rev": "1dd63230066a93c61ab7a66934eb0aae3f1a3613",
        "sha256": "1xn8m62ypg13jh4zf101qmfa6gy8cl923fgwvk9c33m573h3k154",
        "type": "tarball",
        "url": "https://github.com/nmattia/naersk/archive/1dd63230066a93c61ab7a66934eb0aae3f1a3613.tar.gz",
        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
    },
    "niv": {
        "branch": "master",
        "description": "Easy dependency management for Nix projects",
        "homepage": "https://github.com/nmattia/niv",
        "owner": "nmattia",
        "repo": "niv",
        "rev": "f73bf8d584148677b01859677a63191c31911eae",
        "sha256": "0jlmrx633jvqrqlyhlzpvdrnim128gc81q5psz2lpp2af8p8q9qs",
        "type": "tarball",
        "url": "https://github.com/nmattia/niv/archive/f73bf8d584148677b01859677a63191c31911eae.tar.gz",
        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
    },
    "nixpkgs": {
        "branch": "nixos-19.09",
        "description": "A read-only mirror of NixOS/nixpkgs tracking the released channels. Send issues and PRs to",
        "homepage": "https://github.com/NixOS/nixpkgs",
        "owner": "NixOS",
        "repo": "nixpkgs-channels",
        "rev": "e10c65cdb35b6a66491e47e5a85f5d456b4f4eea",
        "sha256": "19csb2s3wyav83zcw9dw488zk2fnz6wcxxz8q6hy43dbph86hxwm",
        "type": "tarball",
        "url": "https://github.com/NixOS/nixpkgs-channels/archive/e10c65cdb35b6a66491e47e5a85f5d456b4f4eea.tar.gz",
        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
    }
}
```

```nix
{pkgs, fetchFromGitHub }:

let
  name = "vector";
  version = "0.9.2";
  sources = import ./nix/sources.nix;
  naersk = pkgs.callPackage sources.naersk { };

in naersk.buildPackage {
  inherit name version;
  src = fetchFromGitHub {
    owner = "timberio";
    repo = "vector";
    rev = "v${version}";
    sha256 = "050x6blm11b90kl2bdlk4v5f43fy5js1fpsykayyx9a3q9vd60cv";
  };
}
```

```
$ nix-build -E '(import <nixpkgs> {}).callPackage ./default.nix {}'  --show-trace
error: while evaluating the attribute 'builtDependencies' of the derivation 'vector-0.9.2' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/build.nix:133:5:
while evaluating the attribute 'cargoconfig' of the derivation 'vector-deps-0.9.2' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/build.nix:133:5:
while evaluating 'toTOML' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:119:12, called from /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/build.nix:145:19:
while evaluating 'concatMapStringsSep' at /nix/store/2pdw4wsfdg0mrg9d6kj6myyimimvnj4d-nixos-20.09pre226148.0f5ce2fac0c/nixos/lib/strings.nix:88:5, called from /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:132:7:
while evaluating 'foldl' at /nix/store/2pdw4wsfdg0mrg9d6kj6myyimimvnj4d-nixos-20.09pre226148.0f5ce2fac0c/nixos/lib/lists.nix:80:20, called from /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:122:14:
while evaluating 'foldl'' at /nix/store/2pdw4wsfdg0mrg9d6kj6myyimimvnj4d-nixos-20.09pre226148.0f5ce2fac0c/nixos/lib/lists.nix:82:16, called from /nix/store/2pdw4wsfdg0mrg9d6kj6myyimimvnj4d-nixos-20.09pre226148.0f5ce2fac0c/nixos/lib/lists.nix:86:8:
while evaluating anonymous function at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:124:16, called from /nix/store/2pdw4wsfdg0mrg9d6kj6myyimimvnj4d-nixos-20.09pre226148.0f5ce2fac0c/nixos/lib/lists.nix:85:14:
while evaluating 'tomlTy' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:102:12, called from /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:126:20:
while evaluating the attribute 'v' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/builtins/to-toml.nix:130:44:
while evaluating the attribute 'source' at /nix/store/g76zggan9y6g82wais1zmrwx5d41vpvd-source/build.nix:146:7:
No 'rev', 'tag' or 'branch' specified
```

### PR DESCRIPTION
**Pull request by yusdacra**
This should fix #109 I believe. I'm not quite sure of the code, especially the `matchByName` function, but it works perfectly on one of my projects and also passes the tests.

### CODE REVIEW

File: lib.nix [hunk 1]
```diff
@@ -79,35 +79,42 @@ rec
...
+                  query = p: p.name == k && (lib.substring 0 (4 + lib.stringLength v.git) p.source) == "git+${v.git}";
+                  extractRevision = url: lib.last (lib.splitString "#" url);
+                  parseLock = lock: rec { inherit (lock) name source; revision = extractRevision source; };
+                  packageLocks = builtins.map parseLock (lib.filter query cargolock.package);
+                  matchByName = lib.findFirst (p: p.name == k) null packageLocks;
+                  matchByRev =
+                    if (isNull (v.rev or null))
+                    then null
+                    else lib.findFirst (p: lib.substring 0 7 p.revision == lib.substring 0 7 v.rev) null packageLocks;
+                  val = if (isNull (v.rev or null)) then v // { rev = matchByRev.revision or matchByName.revision; } else v;
```
Comment (nmattia) [inline lib.nix:99]: Here, why go through `matchByRev`? isn't `matchByRev.revision` going to be equal to `v.rev`?

**Review by nmattia — COMMENTED** [review]
Nice, thank you very much! Just got one question.

File: lib.nix (hunk 1 above)
Comment (yusdacra) [reply lib.nix:99]: Oh yeah, that's right. I only kept that since it was in the original code as well. I can remove that.

Comment (yusdacra) [reply lib.nix]: Alright 10cb38e2704bb1a6cf746f2699da7287ce2e05b0 does that. Tested and it still works correctly on my project.

**Review by nmattia — APPROVED** [review]
Nice, thanks for addressing the comment!

**Pull request merged**